File: OvalensPlanner/packages/shared/src/utils/tax-year.ts
export function getCurrentTaxYear(): string {
  const now = new Date();
  const year = now.getFullYear();
  const month = now.getMonth() + 1;
  const day = now.getDate();

  // Tax year starts April 6
  if (month < 4 || (month === 4 && day < 6)) {
    return `${year - 1}/${(year % 100).toString().padStart(2, "0")}`;
  }
  return `${year}/${((year + 1) % 100).toString().padStart(2, "0")}`;
}

export function daysUntilTaxYearEnd(): number {
  const now = new Date();
  const year = now.getFullYear();
  const month = now.getMonth() + 1;
  const day = now.getDate();

  let endYear = year;
  if (month > 4 || (month === 4 && day >= 6)) {
    endYear = year + 1;
  }

  const taxYearEnd = new Date(endYear, 3, 5); // April 5
  const diffTime = taxYearEnd.getTime() - now.getTime();
  return Math.ceil(diffTime / (1000 * 60 * 60 * 24));
}

File: OvalensPlanner/apps/api/app/utils/tax_year.py
"""Tax year utilities."""

from datetime import date


def get_default_tax_year() -> str:
    """Default tax year for calculations: from DEFAULT_TAX_YEAR config or current date."""
    try:
        from app.config import get_settings
        settings = get_settings()
        if settings.default_tax_year:
            return settings.default_tax_year
    except Exception:
        pass
    return get_current_tax_year()


def get_current_tax_year() -> str:
    """Get the current UK tax year label (e.g., '2025/26')."""
    today = date.today()
    year = today.year
    month = today.month
    day = today.day

    if month < 4 or (month == 4 and day < 6):
        start_year = year - 1
    else:
        start_year = year

    return f"{start_year}/{(start_year + 1) % 100:02d}"


def days_until_tax_year_end() -> int:
    """Calculate days remaining until the current tax year ends (April 5)."""
    today = date.today()
    year = today.year
    month = today.month
    day = today.day

    end_year = year
    if month > 4 or (month == 4 and day >= 6):
        end_year = year + 1

    tax_year_end = date(end_year, 4, 5)
    return (tax_year_end - today).days
